internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,260 @@
+		m.log(fmt.Sprintf("DEBUG: Applying minimum interval delay: %v\n", m.minRequestInterval-timeSinceLastRequest))
+		time.Sleep(m.minRequestInterval - timeSinceLastRequest)
+	}
+
+	if rateLimitInfo != nil && rateLimitInfo.ShouldThrottle(m.threshold) {
+		delay := m.calculateDelay(rateLimitInfo)
+		if delay > 0 {
+			m.log(fmt.Sprintf("DEBUG: Throttling request - delay: %v (remaining: %d, reset: %d, threshold: %.2f)\n",
+				delay, rateLimitInfo.Remaining, rateLimitInfo.Reset, m.threshold))
+			time.Sleep(delay)
+		}
+	} else {
+		if rateLimitInfo == nil {
+			m.log("DEBUG: No rate limit info available yet\n")
+		} else {
+			m.log(fmt.Sprintf("DEBUG: Not throttling - remaining: %d, limit: %d, ratio: %.2f, threshold: %.2f\n",
+				rateLimitInfo.Remaining, rateLimitInfo.Limit,
+				float64(rateLimitInfo.Remaining)/float64(rateLimitInfo.Limit), m.threshold))
+		}
+	}
+
+	m.mu.Lock()
+	m.lastRequestTime = time.Now()
+	m.mu.Unlock()
+
+	m.log("DEBUG: RequestMiddleware completed successfully\n")
+
+	return nil
+}
+
+func (m *Manager) ResponseMiddleware(client *resty.Client, resp *resty.Response) error {
+	m.log(fmt.Sprintf("DEBUG: ResponseMiddleware called, enabled: %v\n", m.enabled))
+
+	if !m.enabled {
+		m.log("DEBUG: Rate limiting disabled - response middleware returning early\n")
+		return nil
+	}
+
+	m.log("DEBUG: Setting up defer function for semaphore release\n")
+
+	defer func() {
+		m.log("DEBUG: Defer function executing - attempting to release semaphore\n")
+
+		select {
+		case <-m.requestSemaphore:
+			m.log("DEBUG: Successfully released semaphore slot\n")
+		case <-time.After(1 * time.Second):
+			m.log("DEBUG: Semaphore release timeout - continuing anyway\n")
+		}
+
+		m.log("DEBUG: Updating active request count\n")
+
+		m.mu.Lock()
+		m.activeRequests--
+		if m.activeRequests < 0 {
+			m.activeRequests = 0 // Prevent negative count
+		}
+		m.log(fmt.Sprintf("DEBUG: Active requests now: %d\n", m.activeRequests))
+		m.mu.Unlock()
+
+		m.log("DEBUG: Defer function completed\n")
+	}()
+
+	m.log("DEBUG: Extracting rate limit headers\n")
+
+	limitHeader := resp.Header().Get("x-ratelimit-limit")
+	periodHeader := resp.Header().Get("x-ratelimit-period")
+	remainingHeader := resp.Header().Get("x-ratelimit-remaining")
+	resetHeader := resp.Header().Get("x-ratelimit-reset")
+
+	m.log(fmt.Sprintf("DEBUG: Rate limit headers - limit: %q, period: %q, remaining: %q, reset: %q\n",
+		limitHeader, periodHeader, remainingHeader, resetHeader))
+
+	if limitHeader != "" && remainingHeader != "" {
+		m.log("DEBUG: Parsing rate limit headers\n")
+
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		rateLimitInfo := &RateLimitInfo{}
+
+		if limit, err := strconv.Atoi(limitHeader); err == nil {
+			rateLimitInfo.Limit = limit
+		}
+		if period, err := strconv.Atoi(periodHeader); err == nil {
+			rateLimitInfo.Period = period
+		}
+		if remaining, err := strconv.Atoi(remainingHeader); err == nil {
+			rateLimitInfo.Remaining = remaining
```

Comment:
Are there edge cases regarding updating the `Remaining` while we have `activeRequests`?

Reply:
Found some interesting edge cases and dealt with them, good point